Add Symmetric Encryption

GaelGuegan opened the pull request:
### Description

Adding Symmetric Encryption : **tpm2-tss-engine-ciphers.c**. (Issue #70)

This is done by handling structure [EVP_CIPHER_meth_new](https://www.openssl.org/docs/manmaster/man3/EVP_CIPHER_meth_new.html).

### **How to use ?** 

1. Shell Command Line utility :
```shell
> Create a persistent sym key
> openssl enc -aes-256-cbc -e -engine tpm2tss -in data.txt -out enc_data -K 81000001 -iv 0123456789012345
> openssl enc -aes-256-cbc -d -engine tpm2tss -in enc_data -out dec_data.txt -K 81000001 -iv 0123456789012345
```

See the shell script : [ciphers.sh](https://github.com/GaelGuegan/tpm2-tss-engine/blob/master/test/ciphers.sh).

2. Openssl API in C code :
```c
strcpy(key , "81000001");
EVP_EncryptInit_ex(ctx, EVP_aes_256_cbc(), libtpm2tss, key, iv);
EVP_EncryptUpdate(ctx, pBufOut, &tmpLen, pBufIn, inLen);
EVP_EncryptFinal_ex(ctx, pBufOut + outLen, &tmpLen);
```

See official [wiki](https://wiki.openssl.org/index.php/EVP_Symmetric_Encryption_and_Decryption).

3. Create Key : 
```shell
tpm2tss-genkey -a aes -m cfb -s 128 sym_key
```

### Problems

For [now](https://github.com/tpm2-software/tpm2-tss-engine/pull/77#issuecomment-466006605), it is only working with **persistent handle**.  We **cannot** passed key file through the openssl app ([Doc](https://www.openssl.org/docs/man1.0.2/man1/openssl-enc.html)), only hex string.

Signed-off-by: Gael Guegan <gael.guegan@non.se.com>

GaelGuegan commented:
Currently checking why CI failed ...

GaelGuegan commented:
After having trouble with sserver.sh, clang, openssl 1.0.2 ...

Finally, CI is passing. 

@AndreasFuchsSIT Can it be merged ?

AndreasFuchsSIT commented:
Glad you could make it work, I know the hassle... ;-)

I'll have a closer look over it after the 1.0 engine release. And queue it up for the 1.1 release...

From a first look, I think this still lacks tpm2tss-genkey support and loading of these, documentation in man-pages and bash-completion.

GaelGuegan commented:
I didn't find a way to use tpm2tss-genkey.  Because of the interface **openssl enc** which does not have options to read a key from a file. 

From the man pages of [openssl](https://www.openssl.org/docs/manmaster/man1/enc.html) :
> **-k password** :  the password to derive the key from. This is for compatibility with previous versions of OpenSSL. Superseded by the -pass argument.
**-kfile filename** :   read the password to derive the key from the first line of filename. This is for compatibility with previous versions of OpenSSL. Superseded by the -pass argument.
**-K key** :  The actual key to use: this must be represented as a string comprised only of hex digits. If only the key is specified, the IV must additionally specified using the -iv option. When both a key and a password are specified, the key given with the -K option will be used and the IV generated from the password will be taken. It does not make much sense to specify both key and password.

I was only able to use the option -K to pass a persistent handle of the TPM. Still looking for solutions ...

AndreasFuchsSIT commented:
Hmmmm..... If `-K` really requires the actual key, then the TPM handle does not suit this IMHO, because it's just the handle...

GaelGuegan commented:
I did that because we were doing the same with RSA. [See this test](https://github.com/tpm2-software/tpm2-tss-engine/blob/933d5d378a83b9b47599e452c703436bc5b4f39b/test/rsasign_persistent.sh#L37) : 
`echo "abc" | openssl pkeyutl -engine tpm2tss -keyform engine -inkey ${HANDLE} -sign -in ${DIR}/mydata.txt -out ${DIR}/mysig -passin stdin`

AndreasFuchsSIT commented:
I know, and I only justify it due to `-keyform engine`. Still no clean, but somewhat arguable.
Give me some time to think about this. Doc and Man would still be good though...

GaelGuegan commented:
I have made a working function tpm2tss_sym_genkey(), and write manual.

I am closing this temporarily, because I need to see with my company about copyright (I know nothing about this kind of stuff, sorry).

Link to my personal repo : https://github.com/GaelGuegan/tpm2-tss-engine

GaelGuegan commented:
CI passed if you click on details, don't know why it is blocked here

GaelGuegan commented:
I have made a local branch that fixed the **bug** when trying to encrypt string of **32 bytes or longer**.

However the CI my test is not passing with openssl version 1.0.2

AndreasFuchsSIT commented on src/tpm2tss-genkey.c:
```diff
@@ -73,13 +74,15 @@ static const struct option long_options[] = {
     TPMI_ALG_PUBLIC alg;
     TPMI_ECC_CURVE curve;
+    TPMI_ALG_SYM_MODE sym_mode;
     int exponent;
```
The help string, the longopts and this struct use different names, i.e. mode, symmode, sym_mode. Those should be unified.

AndreasFuchsSIT commented on src/tpm2tss-genkey.c:
```diff
@@ -151,6 +158,20 @@ parse_opts(int argc, char **argv)
             break;
+        case 'm':
+            if (strcasecmp(optarg, "cfb") == 0) {
+                opt.alg = TPM2_ALG_CFB;
```
Shouldn't this all be opt.sym_mode ?

GaelGuegan replied:
aah yes sorry I have seen this mistake I have already fixed it in my other branch. Either I merge my branch in master (but CI not currently passing for openssl 1.0.2), either I correct it directly in master

GaelGuegan replied:
also there are some other errors when managing sym mode

AndreasFuchsSIT commented on src/tpm2tss-genkey.c:
```diff
@@ -328,6 +378,10 @@ main(int argc, char **argv)
         tpm2Data = genkey_ecdsa();
         break;
+    case TPM2_ALG_AES:
+    case TPM2_ALG_CAMELLIA:
```
You sould add a line with `/* FALLTHROUGH */` here so static code analysis knows it is intentional.

GaelGuegan replied:
OK

GaelGuegan commented:
@AndreasFuchsSIT Do you need something else ? more documentation ? more comments ?

AndreasFuchsSIT commented:
Looks like you are happily working on this.
Give me a heads up when you want me to review it.

GaelGuegan commented:
Okay does not matter I have made a patch anyway for myself, you can take your time

AndreasFuchsSIT commented:
Since 1.0.0 is out now, I'd like to start merging the backloged PRs.
How is it going on your side ? (When) Do you plan to rebase and squash your contributions to current master ?

GaelGuegan commented:
Do I need to rebase from my side, I think I have done it some time ago.

I didn't work on it, since last time.

GaelGuegan commented:
Ah ok I didn't see that there are conflicts, I will rebase.

GaelGuegan commented:
@AndreasFuchsSIT Does it not pose a problem that I was not able to make it works with "tpm2tss-genkey" file ? I can create the file but don't how to get it through openssl. (see https://github.com/tpm2-software/tpm2-tss-engine/pull/77#issuecomment-466006605)

AndreasFuchsSIT commented:
@GaelGuegan you merged but you did not rebase. Could you rebase and squash your contribution such that they are a single commit on top of current master ? That would make things much easier.

Good point about the "cannot read key from file". That is a problem. I guess that actually does still require a solution to be found. :-(

GaelGuegan commented:
I have rebased and squashed as you asked, and CI is passing. 

I did not do it on purpose but there is twice my name in the commit, because of my wrong git configuration sorry.

GaelGuegan commented:
@AndreasFuchsSIT I have rebased and squashed, is that ok for you ?

AndreasFuchsSIT commented:
Yes, that's great...
Now let's figure out how to get the TPM keyids in there...
Maybe we could ask openssl upstream developers if they have something in mind...

GaelGuegan commented:
Currently, I don't think that the **openssl app** can pass file. Maybe that in the meantime that openssl developpers find a solution, if there is one, shouldn't we merge this PR ?

However, I think it would be possible to use our key file through **openssl API** : EVP_Encrypt(...).
see [wiki](https://www.openssl.org/docs/manmaster/man3/EVP_EncryptInit.html) and [man](https://wiki.openssl.org/index.php/EVP_Symmetric_Encryption_and_Decryption).

GaelGuegan commented:
Maybe, one solution would be to get our filename, open and read it in our engine. But the filename need to be a hexadecimal string because of that [function](https://github.com/openssl/openssl/blob/master/apps/enc.c#L647).

AndreasFuchsSIT commented on src/tpm2-tss-engine-ciphers.c:
```diff
@@ -0,0 +1,527 @@
+
+    /* Use blob context */
+    } else {
+        //if (!tpm2tss_tpm2data_read((char *)key, tpm2Data))
```
If we want to merge the current version, this should probably be removed.

GaelGuegan replied:
Previous code : 
```c
    /* Use blob context */
    } else {
        //if (!tpm2tss_tpm2data_read((char *)key, tpm2Data))
        //    return 0;
    }
return 1;
```
Current code :
```c
    /* Use blob context */
    } else {
        return 0;
    }
return 1;
```
Because of that, `tpm2_cipher_init_key()` is failing when doing the `test/sserver.sh` with openssl 1.1.0.

AndreasFuchsSIT commented on src/tpm2-tss-engine-ciphers.c:
```diff
@@ -0,0 +1,527 @@
+}
+
+int
+tpm2tss_sym_genkey(EVP_CIPHER_CTX *cipher, TPMI_ALG_PUBLIC algo,
```
If we only support persistent keys for now, then this function should be removed

AndreasFuchsSIT commented on src/tpm2tss-genkey.c:
```diff
@@ -50,7 +50,8 @@ char *help =
     "    <filename>      storage for the encrypted private key\n"
     "Options:\n"
-    "    -a, --alg       public key algorithm (rsa, ecdsa) (default: rsa)\n"
+    "    -a, --alg       public key algorithm (rsa, ecdsa, aes) (default: rsa)\n"
```
If we only support persistent keys for now, then all changes to this file should be removed.

AndreasFuchsSIT commented on test/ciphers.sh:
```diff
@@ -0,0 +1,60 @@
+tpm2_evictcontrol -a o -c ${HANDLE}
+
+# Test Gen Key
+tpm2tss-genkey -a aes -m cfb -s 128 ${DIR}/mykey
```
This should be removed, since there is no way for us to use the generated key afterwards.

AndreasFuchsSIT commented on src/tpm2-tss-engine-ciphers.c:
```diff
@@ -0,0 +1,527 @@
+        tpm2Data = NULL;
+
+    /* Read persistent key, use for openssl API : EVP_EncryptInit_ex() */
+    } else if (strncmp((char *)key, "0x81", 4) == 0) {
```
Since this is never used, we should just remove it

AndreasFuchsSIT commented on src/tpm2-tss-engine-ciphers.c:
```diff
@@ -0,0 +1,527 @@
+};
+
+static int
+convert_array_hex_to_int(const unsigned char *in, size_t size)
```
This naming is misleading, since it's not a hex array. Rather this is a `big-endian-array-to-native` conversion. My recommended code to also prevent page-align problems would be:
```
if (size != 4) fail;
uint32 r;
memcpy(&r, in, 4);
r = be32toh(r);
```

GaelGuegan replied:
Okay I needed to add the flag : AX_ADD_PREPROC_FLAG([-D_DEFAULT_SOURCE]) in order to use the function be32toh() with <endian.h>

AndreasFuchsSIT commented on man/tpm2tss_sym_genkey.3.md:
```diff
@@ -0,0 +1,38 @@
+% tpm2tss-tpm2data_write(1) tpm2-tss-engine | Library calls
```
This file could probably go if we support persistent only

AndreasFuchsSIT commented on man/tpm2tss-genkey.1.md:
```diff
@@ -24,14 +24,17 @@ key information. This file can then be loaded with OpenSSL using
 # OPTIONS
```
All these changes can go, if we support only persistent

GaelGuegan commented:
CFB and OFB are now the two symmetric modes available. CBC was apparently causing troubles with openssl sserver & sclient. 

However if a mode is already defined within the TPM persistent key, it will use it.

GaelGuegan commented:
:wink:

AndreasFuchsSIT commented on src/tpm2-tss-engine-ciphers.c:
```diff
@@ -0,0 +1,393 @@
+
+    /* Copy in_data : unsigned char* to TPM2B_MAX_BUFFER */
+    in_data = OPENSSL_malloc(sizeof(TPM2B_MAX_BUFFER));
+    if (tpm2Data == NULL) {
```
you mean `in_data` ?

AndreasFuchsSIT commented on src/tpm2-tss-engine-ciphers.c:
```diff
@@ -0,0 +1,393 @@
+    TSS2_RC ret;
+    ESYS_AUXCONTEXT eactx = (ESYS_AUXCONTEXT){ NULL, NULL};
+    ESYS_TR keyHandle = ESYS_TR_NONE;
+    TPM2B_MAX_BUFFER *out_data, *in_data;
```
in_data is never freed.
Also you don't have to malloc it, you could just declare `TPM2_MAX_BUFFER in_data;` on stack instead of a pointer.

AndreasFuchsSIT commented on src/tpm2-tss-engine-ciphers.c:
```diff
@@ -0,0 +1,393 @@
+        ERR(tpm2_do_cipher, ERR_R_MALLOC_FAILURE);
+        goto error;
+    }
+    memcpy(in_data->buffer, in, inl);
```
You should check that `inl < sizeof(in_data->buffer)` before the memcpy...

AndreasFuchsSIT commented on src/tpm2-tss-engine-ciphers.c:
```diff
@@ -0,0 +1,393 @@
+                                &out_data,
+                                &iv_out );
+    if (ret == TPM2_RC_COMMAND_CODE) {
+        DBG("Esys_EncryptDecrypt2 : FAILED\n");
```
I'd add "Falling back to Esys_EncryptDecrypt" or something like that.

AndreasFuchsSIT commented on src/tpm2-tss-engine-ciphers.c:
```diff
@@ -0,0 +1,393 @@
+}
+
+static int
+tpm2_cipher_cleanup(EVP_CIPHER_CTX *ctx)
```
Since this is basically a no-op, can you just not declare this and pass NULL in the struct ?

AndreasFuchsSIT commented on src/tpm2-tss-engine-ciphers.c:
```diff
@@ -0,0 +1,393 @@
+}
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000
+static EVP_CIPHER tpm2_aes_256_ofb =
```
Sorry for being so ignorant, but could you explain the 256 and ofb128 to me ?
I don't know what they mean.
Does that mean that aes 128 is not supported ?

I guess a comment in code would also be good...

GaelGuegan replied:
So this is about the **selector function** and **NIDS**, I guess. So will try to explain my understanding of it.

So at first, you can try to type a `openssl enc -ciphers` on your machine. You will see there the different openssl symmetric encryption available. We have to define which algos of this list , our engine will implement. Currently, I have defined these two : 
```c
static int tpm2_cipher_nids[] = {
    NID_aes_256_cfb128,
    NID_aes_256_ofb128,
    0
};
```
Which are defined by AES algorithm, 256 bits key size and cfb/ofb encryption mode.
When someone call `openssl enc -aes-256-cfb -e -engine tpm2tss ...`, our engine implementation will be used.

Moreover, IDs are defined [here](https://github.com/openssl/openssl/blob/master/include/openssl/obj_mac.h#L2735) in the openssl repo. 

As i said before I took the gost engine as example, we can see [here](https://github.com/gost-engine/engine/blob/master/gost_eng.c#L345) the implementation of its cipher selector. Also another really simple cipher engine exits [here](https://github.com/bigbrett/wsaesengine/blob/master/src/wsaesengine.c#L193).

GaelGuegan replied:
If you have other question, don't hesistate. I will try to be the clearest possible and explain my understanding.

AndreasFuchsSIT commented on src/tpm2-tss-engine-ciphers.c:
```diff
@@ -0,0 +1,393 @@
+    tpm2_cipher_init_key,               // Init key
+    tpm2_do_cipher,                     // Encrypt/Decrypt
+    tpm2_cipher_cleanup,                // Cleanup
+    sizeof(TPM2_DATA),                  // Context size
```
I don't think this is correct. The man-page says `EVP_CIPHER_meth_set_impl_ctx_size() sets the size of the EVP_CIPHER's implementation context so that it can be automatically allocated.` but we are malloc'ing it ourselves. But I also don't know what this is really for and how we should use it.

GaelGuegan replied:
To be honest, I don't fully understand this stuff neither.

I based my code by taking the [gost engine](https://github.com/gost-engine/engine) as example. In there code, we can see that there are doing a `EVP_CIPHER_meth_set_impl_ctx_size(_hidden_magma_cbc, sizeof(struct ossl_gost_cipher_ctx))` [here](https://github.com/gost-engine/engine/blob/master/gost_crypt.c#L248) and then `struct ossl_gost_cipher_ctx *c = EVP_CIPHER_CTX_get_cipher_data(ctx);` [here](https://github.com/gost-engine/engine/blob/master/gost_crypt.c#L461).

So maybe I should put a 0 value here, I tried and it does not seem to change anything. I can change it if you want ?

AndreasFuchsSIT commented on test/ciphers.sh:
```diff
@@ -0,0 +1,56 @@
+
+set -eufx
+
+export OPENSSL_ENGINES=${PWD}/.libs
```
There were some simplification merged for the test scripts in a recent PR.
Please look at the current examples https://github.com/tpm2-software/tpm2-tss-engine/blob/master/test/rsasign.sh

AndreasFuchsSIT commented:
> 
> 
> CFB and OFB are now the two symmetric modes available. CBC was apparently causing troubles with openssl sserver & sclient.
> 
> However if a mode is already defined within the TPM persistent key, it will use it.

I guess that had to do with the comment I posted here: https://github.com/tpm2-software/tpm2-tss-engine/pull/77#discussion-diff-281226821R163

Please see that we enable CBC mode for the TPM and still have sserver running. It's basically the counter test that the engine does not interfere with any other operations; especially if the engine shall only be used for asymmetric operations.
Also I don't know about the switch-case statement in the selector function. Maybe the default there also needs to be a standard engine.

AndreasFuchsSIT commented:
As a final comment: Since only very few TPMs support symmetric operations, I'd like to wrap support for it in a configure parameter `--enable-symmetric`; just to minimize interference possibilities with the asymmetric operations.

GaelGuegan commented on src/tpm2-tss-engine-ciphers.c:
```diff
@@ -0,0 +1,393 @@
+        *tpm2Data = NULL;
+
+    /* Read persistent key, use for openssl app : openssl enc */
+    } else if ((key[0] == 0x81) && (key[HANDLE_SIZE/2 + 1] == 0)) {
```
That not possible here. `strnlen(key, HANDLE_SIZE*2+1)` will always return 1 because the second byte of data is `0x00` in a key handle, for example `0x81000001`

AndreasFuchsSIT replied:
Oh sorry, of course you are right.

GaelGuegan commented on src/tpm2-tss-engine-ciphers.c:
```diff
@@ -0,0 +1,393 @@
+    /* Get App Data */
+    tpm2Data = EVP_CIPHER_CTX_get_app_data(ctx);
+    if (tpm2Data == NULL || tpm2Data->pub.size == 0 || inl == 0) {
+        return 0;
```
mmmh don't know of to do that here.

And if tpm2Data is NULL, tpm2_cipher_init_key() function should has failed before.

GaelGuegan commented:
I have done some changes that you requested. However can you relaunch my CI build ? a weird error appeared, CI was not able to clone openssl.

AndreasFuchsSIT commented:
You can't but I did

AndreasFuchsSIT commented:
Since https://github.com/tpm2-software/tpm2-tools/pull/1476 I'll retrigger the CI...

GaelGuegan commented:
don't understand why CI is not passing here, because it passed in my personnal branch [here](https://travis-ci.org/GaelGuegan/tpm2-tss-engine)

AndreasFuchsSIT commented:
Looks very much stalled. So closing
